include/seqan3/search/algorithm/search.hpp

```diff
@@ -87,63 +168,29 @@ inline auto search(queries_t && queries,
-        static_assert(std::ranges::sized_range<value_type_t<queries_t>>,
-                      "Elements of the query collection must model sized_range.");
-    }
+    using search_traits_t = detail::search_traits<configuration_t>;
 
-    static_assert(detail::is_type_specialisation_of_v<remove_cvref_t<configuration_t>, configuration>,
-                  "cfg must be a specialisation of seqan3::configuration.");
-
-    using cfg_t = remove_cvref_t<configuration_t>;
-
-    if constexpr (cfg_t::template exists<search_cfg::max_error>())
-    {
-        auto & [total, subs, ins, del] = get<search_cfg::max_error>(cfg).value;
-        if (subs > total)
-            throw std::invalid_argument("The substitution error threshold is higher than the total error threshold.");
-        if (ins > total)
-            throw std::invalid_argument("The insertion error threshold is higher than the total error threshold.");
-        if (del > total)
-            throw std::invalid_argument("The deletion error threshold is higher than the total error threshold.");
-    }
-    else if constexpr (cfg_t::template exists<search_cfg::max_error_rate>())
+    // If no mode was set, default to search all hits.
+    if constexpr (!search_traits_t::has_mode_configuration)
     {
-        auto & [total, subs, ins, del] = get<search_cfg::max_error_rate>(cfg).value;
-        if (subs > total)
-            throw std::invalid_argument("The substitution error threshold is higher than the total error threshold.");
-        if (ins > total)
-            throw std::invalid_argument("The insertion error threshold is higher than the total error threshold.");
-        if (del > total)
-            throw std::invalid_argument("The deletion error threshold is higher than the total error threshold.");
-    }
-
-    if constexpr (cfg_t::template exists<search_cfg::mode>())
-    {
-        if constexpr (cfg_t::template exists<search_cfg::output>())
-            return detail::search_all(index, queries, cfg);
-        else
-            return detail::search_all(index, queries, cfg | search_cfg::output{search_cfg::text_position});
+        return search(std::forward<queries_t>(queries), index, cfg | search_cfg::mode{search_cfg::all});
     }
     else
-    {
-        configuration const cfg2 = cfg | search_cfg::mode{search_cfg::all};
-        if constexpr (cfg_t::template exists<search_cfg::output>())
-            return detail::search_all(index, queries, cfg2);
+    {   // If no output configuration was, default to returning text positions.
+        if constexpr (!search_traits_t::has_output_configuration)
+        {
+            return search(std::forward<queries_t>(queries), index, cfg | search_cfg::output{search_cfg::text_position});
+        }
         else
-            return detail::search_all(index, queries, cfg2 | search_cfg::output{search_cfg::text_position});
+        {
+            detail::search_configuration_validator::validate_query_type<queries_t>();
```

Comment:
> You only validate the query type if a mode and output config was set but shouldn't this be always checked? 

No matter what you specify in the config it will always pass through this lines of code. If some options are not present a default configuration for them will be added and the function is called again with the modified config.
So I think that should be fine?

Reply:
Ah I did not see that, that's not so easy to spot. I would prefer a lambda that just alters the config :grin:  but yes it's fine like this.